File: generated_code/GeminiPro/Python/005_SplayTree.py
import sys

class SplayTree:
    """
    A complete, self-contained implementation of a Splay Tree.

    This class implements a dictionary-like set for storing unique integer keys.
    It supports insert, delete, and search operations. The key feature of a
    splay tree is that any accessed node (or its parent if the node is not found)
    is moved to the root of thetree through a series of rotations, a process
    called "splaying". This ensures that frequently accessed elements are
    kept near the top of the tree, providing good amortized performance.

    The main methods are:
    - insert(key): Adds a key to the tree.
    - delete(key): Removes a key from the tree.
    - search(key): Checks for a key's existence and splays the accessed node.
    """

    class _Node:
        """A private helper class representing a node in the Splay Tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation around node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
            if not x.parent:
                self.root = y
            elif x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
            y.left = x
            x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation around node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
            if not x.parent:
                self.root = y
            elif x == x.parent.right:
                x.parent.right = y
            else:
                x.parent.left = y
            y.right = x
            x.parent = y

    def _splay(self, node):
        """
        Performs the splaying operation on a node, moving it to the root.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:  # Zig step
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zag-Zag
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-Zag
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zag-Zig
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def insert(self, key):
        """
        Inserts a key into the Splay Tree.

        If the key already exists, the node with that key is splayed to the root.
        If the key does not exist, it is inserted and the new node becomes the root.

        Args:
            key (int): The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the closest node to the key's insertion point
        self.search(key)

        # After splaying, the root is the closest node.
        # If the key is already present, we're done.
        if self.root.key == key:
            return

        # Otherwise, insert the new key and make it the new root
        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            if self.root.left:
                self.root.left.parent = new_node
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            if self.root.right:
                self.root.right.parent = new_node
            self.root.right = None

        self.root.parent = new_node
        self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the Splay Tree.

        If the key is found, its node is removed, and its parent is splayed
        to the root. If the key is not found, the last accessed node during
        the search is splayed.

        Args:
            key (int): The integer key to delete.
        """
        if not self.root:
            return

        # Splay the node with the given key (or its would-be parent) to the root
        self.search(key)

        # If key is not in the tree, self.root.key will not match
        if self.root.key != key:
            return

        # Now, the node to be deleted is the root.
        # We need to join its left and right subtrees.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # Promote the right subtree to be the new root
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Find the maximum element in the left subtree
            left_subtree.parent = None
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # Splay this max element to the top of the left subtree
            # This makes it the new root of the merged tree
            self.root = left_subtree # Temporarily set root for splay
            self._splay(max_node)

            # Join the original right subtree
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

    def search(self, key):
        """
        Searches for a key in the Splay Tree and splays the accessed node.

        If the key is found, the node containing the key is splayed to the root.
        If the key is not found, the last node visited before reaching a NULL
        pointer is splayed to the root.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = None
        while current:
            last_visited = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right

        # Key not found, splay the last visited node
        if last_visited:
            self._splay(last_visited)
        return False

File: generated_code/GeminiPro/Java/005_SplayTree.java
/**
 * A complete, self-contained implementation of an integer set using a Splay Tree.
 * A Splay Tree is a self-balancing binary search tree with the additional property
 * that recently accessed elements are quick to access again. It performs basic
 * operations such as insertion, look-up and removal in O(log n) amortized time.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree. This is a private inner class
     * as it's an implementation detail of the SplayTree.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * Splays the node with the given key to the root of the tree.
     * If the key is not in the tree, the last accessed node on the search path
     * is splayed to the root. This is a top-down splay implementation.
     *
     * @param root The root of the tree/subtree to splay.
     * @param key The key to splay on.
     * @return The new root of the tree after splaying.
     */
    private Node splay(Node root, int key) {
        if (root == null || root.key == key) {
            return root;
        }

        // Use a dummy node to simplify linking of left and right subtrees.
        Node dummy = new Node(0);
        Node leftTreeMax = dummy;
        Node rightTreeMin = dummy;
        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) {
                    break;
                }
                // Zig-Zig (Right-Right) case
                if (key < current.left.key) {
                    current = rightRotate(current);
                    if (current.left == null) {
                        break;
                    }
                }
                // Link the current root to the right tree
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    break;
                }
                // Zig-Zig (Left-Left) case
                if (key > current.right.key) {
                    current = leftRotate(current);
                    if (current.right == null) {
                        break;
                    }
                }
                // Link the current root to the left tree
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else {
                // Key found
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = dummy.right;
        current.right = dummy.left;

        return current;
    }

    /**
     * Searches for a key in the tree.
     * This operation performs a splay on the accessed node (or the last node on the search path),
     * moving it to the root, as required by the splay tree property.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree.
     * If the key already exists, the tree is splayed on that key, but no new node is added.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree on the key. This brings the closest node to the root.
        root = splay(root, key);

        // If the key already exists, do nothing (as this is a set)
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree on the key. If the key is present, it becomes the root.
        root = splay(root, key);

        // If the key is not in the tree, the root's key won't match. Do nothing.
        if (root.key != key) {
            return;
        }

        // The node to be deleted is now the root. We need to merge its subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to its root.
            // We can splay on the key being deleted, as it's guaranteed to be
            // larger than any key in the left subtree.
            leftSubtree = splay(leftSubtree, key);

            // After splaying, the new root of the left subtree (the max element)
            // has no right child. We can attach the original right subtree there.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}